src/runtime/thread_pool.cc

```diff
@@ -127,99 +130,138 @@ class ParallelLauncher {
   std::vector<std::string> par_errors_;
 };
 
-/*! \brief Working queue for each thread */
-class ParallelTaskQueue {
+/*! \brief Lock-free single-producer-single-consumer queue for each thread */
+class SpscTaskQueue {
  public:
   /*! \brief The task entry */
   struct Task {
     ParallelLauncher* launcher;
     int32_t task_id;
   };
-  ParallelTaskQueue() {
-    ring_.resize(2);
+
+  explicit SpscTaskQueue(char size) :
+    size_(size),
+    buffer_(reinterpret_cast<Task*>(new aligned_t[size_])),
+    head_(0),
+    tail_(0) {
+  }
+
+  ~SpscTaskQueue() {
+    delete[] buffer_;
   }
+
   /*!
-   * \brief Signal to kill the job.
+   * \brief Lock-free enqueue.
+   * \param input The task to be enqueued.
+   * \return Whether the task is enqueued.
    */
-  void SignalForKill() {
-    std::lock_guard<std::mutex> lock(mutex_);
-    exit_now_.store(true);
-    cv_.notify_all();
+  bool Enqueue(const Task& input) {
+    const uint8_t tail = tail_.load(std::memory_order_relaxed);
```

Comment:
use uint32_t or int32_t, since uint8 is likely going to be even more expensive most machine have 32 bits as a word

Reply:
Move Enqueue to a private function, since it is not suppose to be used